### PR TITLE
Add PSR4 autoload for tests

### PR DESCRIPTION
**Pull request by lucasmichot**
Add PSR4 autoload for tests

---

Replace #3556

### CODE REVIEW

File: tests/TestCase.php [hunk 1]
```diff
@@ -1,6 +1,11 @@
 <?php
 
-class TestCase extends Illuminate\Foundation\Testing\TestCase
+namespace App\Tests;
+
+use Illuminate\Contracts\Console\Kernel;
```
Comment (shehi) [inline tests/TestCase.php]: Add alias of ```as ConsoleKernel``` maybe? There are so many "kernels" in this package... Just a thought.

Comment (lucasmichot) [inline tests/TestCase.php]: True, but there's only one in this class

**Comment by skyrpex** [comment]
I've done this always in my Laravel projects... Would love to see it merged! :+1:

**Comment by vrkansagara** [comment]
:+1:  good thought.

**Comment by hernandev** [comment]
Not sure about that. i like to keep the tests on the same namespace than the actual code. it helps a lot with keeping less code (less use statements)

By Default, PHPUnit would recognise the files recursively, following the name structure would be nice but autoloading the files is completely unnecessary.

For having a default namespace on the tests that follows the application namespace: :+1: 
For having tests on the autoload: :-1:

**Comment by shehi** [comment]
@hernandev : You can't remove tests from autoload: ```\App``` namespace is looking for ```/app``` folder by default. ```/tests``` folder however is outside that folder, but as a namespace is inside ```\App``` namespace. It's duality so we have to declare a new autoloader resolution implicitly for new namespace of ```\App\Tests```. 

And PHPUnit is fed an autoloader file from application bootstrap, which technically works via composer's autoloader. Not to mention many IDE's, including PHPStorm need an autoload file location to perform in-IDE-tests as well. So overall, all this needs full compliance with composer. If ```composer dump``` command doesn't recognize all test files, we will have a problem.

What you desire is impossible to achieve I am afraid.

**Comment by hernandev** [comment]
@shehi 

1) I'm against using App\Tests, just App is easier to import classes.
2) Any namespace you put your tests under, will be recognised by PHPUnit. The bootstrap is for PHPUnit to load files OUTSIDE the tests directory.
it's the <testsuites> that loads tests. not PSR-4 Loader.
3) About IDE, no idea, i always run stuff on the command line.

You have some valid points. but i don't think this way would cover most of developers.

I'll keep my vote to same tests and app namespace, no PSR-4. It would not hurt me, but i don't honestly see a point of autoloading tests since you're not instantiating them directly.

Let's see what the other folks have to say.

**Comment by hernandev** [comment]
P.S. About the "duality", if you don't declare the tests under PSR-4, it works fine on same namespace and makes tests shorter, easier. That's my point on using the same namespace.

**Comment by hernandev** [comment]
Sent PR #3563 as PoC that PSR-4 is not necessary and that Tests and App can, indeed, have the same namespace.

**Comment by Zizaco** [comment]
`App\Tests` namsepace seems wierd. I don't like having separate test namespaces.

@lucasmichot see http://stackoverflow.com/a/12141610/1063081

**Comment by rkgrep** [comment]
I think, this discussion leads to nowhere. There are different pros and cons for both versions, and framework does not force you to use specific method. You can override it as you want just as you override most of initial project configuration.

**Comment by lucasmichot** [comment]
@taylorotwell @GrahamCampbell @crynobone @JosephSilber thoughts?

**Comment by crynobone** [comment]
> I think, this discussion leads to nowhere. There are different pros and cons for both versions, and framework does not force you to use specific method. You can override it as you want just as you override most of initial project configuration.

Totally agree with this, regardless which option we choose there will always be a number of users that would prefer the alternative options. So why bother changing the default.

**Comment by GrahamCampbell** [comment]
> So why bother changing the default.

I totally agree. The default we have is fine, and changing it is very trivial.

**Comment by shehi** [comment]
@Zizaco : I disagree. Just another unnecessary ideology from perfectionists. 

And I tend to agree with @rkgrep , some guys here appear to talk for the sake of talking, instead of moving these discussions forward. We are not developing nuclear reactor monitoring tool for God's sake.

Overall, either way is fine: both default and the current PR state. Guys up top, please either decline or merge it and let's be done with it. Thanks @lucasmichot again for this PR and your time.

**Pull request closed without merging**